src/librustc/mir/interpret/memory.rs

```diff
@@ -553,7 +553,10 @@ impl<'a, 'tcx, M: Machine<'tcx>> Memory<'a, 'tcx, M> {
         );
         self.check_bounds(ptr.offset(len, &*self)?, true)?; // if ptr.offset is in bounds, then so is ptr (because offset checks for overflow)
 
-        let locks = self.locks.get_mut(&ptr.alloc_id.0).ok_or(EvalErrorKind::DanglingPointerDeref)?;
+        let locks = match self.locks.get_mut(&ptr.alloc_id.0) {
+            Some(locks) => locks,
+            None => return Ok(()),
+        };
```

Comment:
This doesn't look right -- early-returning here will mean that we don't actually acquire any locks!

Reply:
Yea, but we shouldn't be locking static/constant memory anyway...

Reply:
This will affect all locking, not just CTFE, right?

Also, locking such memory shouldn't harm, an adding extra exception for the locking just makes the code more complicated.

Reply:
It will make the code more complicated if we have to generate the lock structures for constants and statics. I just tried it and it requires quite some additional logic.

You need to remember that statics and constants are now cached in the `TyCtxt`, but we can't "cache" the locks, since they are modified all the time. If we just place the locks in the normal locking structure, we'll need to add code that looks on the next statement to figure out if it accesses a static/const that doesn't have a lock structure yet and add it to the structure.

Note that `static mut` or statics with interior mutability are not affected. Only constant memory.

I don't see the issue with just ignoring all permanently immutable memory in the locking scheme, but maybe this has an effect on the other locks? I don't know your memory checking well enough to really make a useful comment. Can you elaborate?

Reply:
> Note that static mut or statics with interior mutability are not affected. Only constant memory.

I guess then not having locking is fine; I take it that writes to this memory are raising exceptions?

> I don't see the issue with just ignoring all permanently immutable memory in the locking scheme, but maybe this has an effect on the other locks?

Just looking at this code, locks are ignored if there is no entry in the hash table. So there is a guarantee that mutable memory has an entry in that table? If that's the case, this code is fine; but I think that warrants a comment.

Reply:
> I take it that writes to this memory are raising exceptions?

Yes

> So there is a guarantee that mutable memory has an entry in that table?

Yes

> I think that warrants a comment.

seems reasonable